src/macaron/slsa_analyzer/provenance/intoto/v1/__init__.py

```diff
@@ -12,3 +40,198 @@ class InTotoV1Statement(TypedDict):
+            "The value of attribute '_type' in the in-toto statement must be: 'https://in-toto.io/Statement/v1'",
+        )
+
+    subjects_payload = payload.get("subject")
+    if subjects_payload is None:
+        raise ValidateInTotoPayloadError(
+            "The attribute 'subject' of the in-toto statement is missing.",
+        )
+    if not isinstance(subjects_payload, list):
+        raise ValidateInTotoPayloadError(
+            "The value of attribute 'subject' in the in-toto statement is invalid: expecting a list.",
+        )
+
+    for subject_json in subjects_payload:
+        validate_intoto_subject(subject_json)
+
+    predicate_type = payload.get("predicateType")
+    if predicate_type is None:
+        raise ValidateInTotoPayloadError(
+            "The attribute 'predicateType' of the in-toto statement is missing.",
+        )
+
+    if not isinstance(predicate_type, str):
+        raise ValidateInTotoPayloadError(
+            "The value of attribute 'predicateType' in the in-toto statement is invalid: expecting a string."
+        )
+
+    predicate = payload.get("predicate")
+    if predicate is not None and not isinstance(predicate, dict):
+        raise ValidateInTotoPayloadError(
+            "The value attribute 'predicate' in the in-toto statement is invalid: expecting an object.",
+        )
+
+    return True
+
+
+def validate_intoto_subject(subject: JsonType) -> TypeGuard[InTotoV1ResourceDescriptor]:
+    """Validate a single subject in the in-toto statement.
+
+    See specification: https://github.com/in-toto/attestation/blob/main/spec/v1/resource_descriptor.md
+
+    Parameters
+    ----------
+    subject : JsonType
+        The JSON element representing a single subject.
+
+    Returns
+    -------
+    TypeGuard[InTotoSubject]
+        ``True`` if the subject element is valid, in which case its type is narrowed to an
+        ``InTotoSubject``; ``False`` otherwise.
+
+    Raises
+    ------
+    ValidateInTotoPayloadError
+        When the payload does not follow the expecting schema.
+    """
+    if not isinstance(subject, dict):
+        raise ValidateInTotoPayloadError(
+            "A subject in the in-toto statement is invalid: expecting an object.",
+        )
+
+    name_valid = _validate_property(subject, "name", False, lambda x: isinstance(x, str))
+    if not name_valid:
+        raise ValidateInTotoPayloadError("The subject name is not valid.")
+
+    # At least one of 'uri', 'digest', and 'content' must be valid and present.
+    uri_valid = _validate_property(subject, "uri", True, lambda x: isinstance(x, str))
+    content_valid = _validate_property(subject, "content", True, lambda x: isinstance(x, str))
+    digest_valid = _validate_property(subject, "digest", True, is_valid_digest_set)
+    if not (uri_valid or content_valid or digest_valid):
+        raise ValidateInTotoPayloadError(
+            "One of 'uri', 'digest', or 'content' must be present and valid within 'subject'."
+        )
```

Comment:
Combining different issues into a single error message may not be the best idea since it might be hard to tell what the exact issue is when one comes up.

Since you pass the name of the attribute into the function, it may be better to raise the error within the `_validate_property` function itself with a more specific context and error message, rather than returning a `bool` value, only just to convert it to an exception later. All `if` statements like this one can be removed in that case.